graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -82,84 +51,73 @@ def merge_schemas(schemas_dict):
 
             elif isinstance(new_definition, ast_types.DirectiveDefinition):
-                if new_name in directives:  # existing directive
-                    if new_definition == directives[new_name]:  # definitions agree
-                        continue
-                    else:  # definitions disagree
-                        raise SchemaNameConflictError(
-                            u'Directive "{}" with definition "{}" has already been defined with '
-                            u'definition "{}".'.format(new_name, print_ast(new_definition),
-                                                       print_ast(directives[new_name]))
-                        )
-                # new directive
-                merged_schema_ast.definitions.append(new_definition)  # Add to AST
-                directives[new_name] = new_definition
+                _process_directive_definition(
+                    new_definition, directives, merged_schema_ast
+                )
 
             elif isinstance(new_definition, ast_types.ScalarTypeDefinition):
-                if new_name in scalars:  # existing scalar
-                    continue
-                if new_name in name_to_schema_id:  # new scalar clashing with existing type
-                    raise SchemaNameConflictError(
-                        u'New scalar "{}" clashes with existing type.'.format(new_name)
-                    )
-                # new, valid scalar
-                merged_schema_ast.definitions.append(new_definition)  # Add to AST
-                scalars.add(new_name)
-
-            else:  # Generic type definition
-                if new_name in scalars:
-                    raise SchemaNameConflictError(
-                        u'New type "{}" clashes with existing scalar.'.format(new_name)
-                    )
-                if new_name in name_to_schema_id:
-                    raise SchemaNameConflictError(
-                        u'New type "{}" clashes with existing type.'.format(new_name)
-                    )
-                merged_schema_ast.definitions.append(new_definition)
-                name_to_schema_id[new_name] = cur_schema_id
+                _process_scalar_definition(
+                    new_definition, scalars, six.iterkeys(name_to_schema_id), merged_schema_ast
+                )
+
+            elif isinstance(new_definition, (
+                ast_types.EnumTypeDefinition,
+                ast_types.InterfaceTypeDefinition,
+                ast_types.ObjectTypeDefinition,
+                ast_types.UnionTypeDefinition,
+            )):
+                _process_generic_type_definition(
+                    new_definition, current_schema_id, scalars, name_to_schema_id, merged_schema_ast
+                )
+
+            else:  # All definition types should've been covered
+                raise AssertionError(
+                    u'Missed definition type: "{}"'.format(type(new_definition).__name__)
```

Comment:
I think it is nearly strictly better to just do `.format(new_definition)`. If the `__str__` method is overridden then stringifying the element is likely to produce useful information. If it is not overridden, then `__str__` defaults to `__repr__` which almost definitely produces at least the same amount of information as `type(new_definition).__name__`, (the default implementation includes the type name). Furthermore, `__repr__` method is often overridden by directive defintions so it'd be a good idea to use it.